src/pyo3_extensions/chrono_tz.rs

```diff
@@ -0,0 +1,232 @@
+    pub fn checked_add_signed(&self, dur: chrono::Duration) -> Option<Self> {
+        self.0.checked_add_signed(dur).map(Self)
+    }
+}
+
+impl std::ops::Sub<&ChronoDateTime> for &ChronoDateTime {
+    type Output = Duration;
+
+    #[inline]
+    fn sub(self, rhs: &ChronoDateTime) -> Duration {
+        self.0 - rhs.0
+    }
+}
+
+impl std::ops::Sub<ChronoDateTime> for ChronoDateTime {
+    type Output = Duration;
+
+    #[inline]
+    fn sub(self, rhs: ChronoDateTime) -> Duration {
+        self.0 - rhs.0
+    }
+}
+
+impl ToPyObject for ChronoDateTime {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        let date = self.0.naive_local().date();
+        let time = self.0.naive_local().time();
+        let yy = date.year();
+        let mm = date.month() as u8;
+        let dd = date.day() as u8;
+        let h = time.hour() as u8;
+        let m = time.minute() as u8;
+        let s = time.second() as u8;
+        let ns = time.nanosecond();
+        let (ms, fold) = match ns.checked_sub(1_000_000_000) {
+            Some(ns) => (ns / 1000, true),
+            None => (ns / 1000, false),
+        };
+        // If the python we are using has the zoneinfo package, we can use
+        // the timezone name, otherwise we use fixed offset.
+        let mut tz = None;
+        if let Ok(zoneinfo) = py.import("zoneinfo") {
+            if let Ok(zi) = zoneinfo.getattr("ZoneInfo") {
+                tz = Some(zi.call1((self.0.timezone().name(),)).unwrap().to_object(py));
+            }
+        }
+        let tz = tz.unwrap_or_else(|| self.0.offset().fix().to_object(py));
+        let tz = tz.cast_as(py).unwrap();
+        let datetime = PyDateTime::new_with_fold(py, yy, mm, dd, h, m, s, ms, Some(tz), fold)
+            .expect("Failed to construct datetime");
+        datetime.into()
+    }
+}
+
+impl IntoPy<PyObject> for ChronoDateTime {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        ToPyObject::to_object(&self, py)
+    }
+}
+
+impl FromPyObject<'_> for ChronoDateTime {
+    fn extract(ob: &PyAny) -> PyResult<ChronoDateTime> {
+        let mut dt = <PyDateTime as PyTryFrom>::try_from(ob)?;
+        let tz: chrono_tz::Tz = if let Some(tzinfo) = dt.get_tzinfo() {
+            // If we have a "key" (name) for the timezone, we try to parse it with
+            // chrono_tz. Otherwise we ask python to convert the datetime to UTC
+            // and return UTC as the timezone.
+            if let Ok(key) = tzinfo.getattr("key") {
+                let key: String = key.extract()?;
+                key.parse()
+                    .map_err(|_err| PyTypeError::new_err("Timezone not found"))?
+            } else {
+                let utc = ob
+                    .py()
+                    .import("datetime")?
+                    .getattr("timezone")?
+                    .getattr("utc")?;
+                dt = dt.getattr("astimezone")?.call1((utc,))?.cast_as()?;
+                chrono_tz::Tz::UTC
```

Comment:
What's your thinking behind doing the conversion to UTC and not trying to create a `chrono::offset::FixedOffset` in this case? It feels like it would be nice to round-trip a TZ datetime without this conversion.